hardware_integration/src/can_hardware_interface.cpp

```diff
@@ -16,48 +17,160 @@
+				start_threads();
+#endif
+				retVal = true;
+			}
+		}
+		return retVal;
+	}
+
+	bool CANHardwareInterface::CANHardware::stop()
+	{
+#if !defined CAN_STACK_DISABLE_THREADS && !defined ARDUINO
+		stop_threads();
+#endif
+		if (nullptr != frameHandler)
+		{
+			frameHandler = nullptr;
+		}
+		messagesToBeTransmittedQueue.clear();
+		receivedMessagesQueue.clear();
+		return false;
+	}
+
+	bool CANHardwareInterface::CANHardware::transmit_can_frame(const CANMessageFrame &frame) const
+	{
+		if ((nullptr != frameHandler) && frameHandler->get_is_valid() && frameHandler->write_frame(frame))
+		{
+			return true;
+		}
+		return false;
+	}
+
+	bool CANHardwareInterface::CANHardware::receive_can_frame()
+	{
+		if ((nullptr != frameHandler) && frameHandler->get_is_valid() && (!receivedMessagesQueue.is_full()))
+		{
+			CANMessageFrame frame;
+			if (frameHandler->read_frame(frame))
+			{
+				receivedMessagesQueue.push(frame);
+				return true; // Indicate that a frame was read
+			}
+		}
+		return false;
+	}
+
+#if !defined CAN_STACK_DISABLE_THREADS && !defined ARDUINO
 	CANHardwareInterface::~CANHardwareInterface()
 	{
 		stop_threads();
 	}
 
+	void CANHardwareInterface::CANHardware::start_threads()
+	{
+		receiveThreadRunning = true;
+		if (nullptr == receiveMessageThread)
+		{
+			receiveMessageThread.reset(new std::thread([this]() { receive_thread_function(); }));
+		}
+	}
+
+	void CANHardwareInterface::CANHardware::stop_threads()
+	{
+		receiveThreadRunning = false;
+		if (nullptr != receiveMessageThread)
+		{
+			if (receiveMessageThread->joinable())
+			{
+				receiveMessageThread->join();
+			}
+			receiveMessageThread = nullptr;
+		}
+	}
+
+	void CANHardwareInterface::CANHardware::receive_thread_function()
+	{
+		while (receiveThreadRunning)
+		{
+			if ((nullptr != frameHandler) && frameHandler->get_is_valid())
+			{
+				if (!receive_can_frame())
+				{
+					// There was no frame to receive, so if any other thread wants to do something, let it.
+					std::this_thread::yield();
+				}
+				else
+				{
+					CANHardwareInterface::updateThreadWakeupCondition.notify_all();
+				}
+			}
+			else
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // Arbitrary, but don't want to infinite loop on the validity check.
+			}
+		}
+	}
+#endif
+
 	bool send_can_message_frame_to_hardware(const CANMessageFrame &frame)
 	{
 		return CANHardwareInterface::transmit_can_frame(frame);
 	}
 
-	bool CANHardwareInterface::set_number_of_can_channels(std::uint8_t value)
+	bool CANHardwareInterface::set_number_of_can_channels(std::uint8_t value, std::size_t queueCapacity)
 	{
-		std::lock_guard<std::mutex> lock(hardwareChannelsMutex);
+		LOCK_GUARD(Mutex, hardwareChannelsMutex);
 
-		if (threadsStarted)
+		if (started)
 		{
 			LOG_ERROR("[HardwareInterface] Cannot set number of channels after interface is started.");
 			return false;
 		}
 
 		while (value > hardwareChannels.size())
```

Comment:
While we're in this file we might want to clean up these implicit casts between uint8_t and size_t